lib/EmitVerilog/EmitVerilog.cpp

```diff
@@ -469,7 +483,7 @@ StringRef ModuleEmitter::addName(Value value, StringRef name) {
   }
 
   auto isValidVerilogCharacter = [](char ch) -> bool {
-    return isalpha(ch) || isdigit(ch) || ch == '_';
+    return isalpha(ch) || isdigit(ch) || ch == '_' || ch == '.';
```

Comment:
Maybe it's just me but I think allowing '.' in names could be confusing. (Is it even valid?)

Reply:
I was looking at an example like this using modports:

```
module top;
  i2 i();
  s u1(.i(i.send));
  r u2(.i(i.recv));
endmodule
```

`i.send` and `i.recv` would be the names for the value of the `sv.modport.get` in this case.

Reply:
Ah. I thought this was for the port name rather than the signal name use in the module instantiation. In this case (inline printing), I don't think you'd need to reserve the full `i.send` just `i`. I *think* that `addName` is intended exclusively to make sure there aren't any collisions in declared signals and instance identifiers. In the case where you're declaring the modport out-of-line (if that's legal in SV), you'd have to register the identifier (via `addName`). In the latter case, I don't think we should use `i.send` as the identifier -- I'd rather `i_send` or something like that.

Reply:
>  I thought this was for the port name rather than the signal name use in the module instantiation. 

Technically it is for the modport name used in module instantiation. The emitter for `rtl.instance` is set up to use `getName` to [look up](https://github.com/llvm/circt/blob/main/lib/EmitVerilog/EmitVerilog.cpp#L1776) a name that has already been added using `addName`.

To also answer your comment [here](https://github.com/llvm/circt/pull/302/files#r537716227), that is why I've added the entire interface instance name and modport name to the identifier. The mapping is from Value -> StringRef, so in this case, any uses of a `sv.modport.get` operation will emit the string `<interface name>.<modport name>`.

> In the latter case, I don't think we should use i.send as the identifier -- I'd rather i_send or something like that.

I'm not really sure how to emit something else, am I missing something? The interface instance in this example is named `i`, and the modports are named `send` and `receive`. When the modports are used in module instantiation, shouldn't we emit `i.send` and `i.recv`? I don't see what `i_send` would refer to in your example.

Reply:
> I'm not really sure how to emit something else, am I missing something? The interface instance in this example is named i, and the modports are named send and receive. When the modports are used in module instantiation, shouldn't we emit i.send and i.recv? I don't see what i_send would refer to in your example.

By 'out-of-line', I was thinking somethink like:
```
i2 i();
i.send i_send = i2.send;
s u1(.i(i_send));
```
That's not valid verilog (and I don't even know if there's any valid verilog to do that) but I don't think we should do that. 

> Technically it is for the modport name used in module instantiation. The emitter for rtl.instance is set up to use getName to look up a name that has already been added using addName.

Yep -- I get it now!  I didn't remember that `addName` is also used for that.

I think that allowing '.' breaks signal name re-naming in the case where a `Value` has a name attribute with a '.' in it. (And it'll end up being declared as a `wire`. I'm pretty sure this function is responsible for replacing that character with something which is valid in identifiers so we could end up with `wire foo.t = ...`. Please correct me if I'm wrong.

Reply:
> we could end up with `wire foo.t = ...`

I see... I'm not sure the best way to avoid that behavior and still allow names like `i.send` for interface modports. I guess I could remove `.` from the set of valid characters, but add some more special casing to allow it when the value is from an interface modport.

Reply:
Maybe you could not assign a name and figure out some way to output that string directly in the expression output code?

Reply:
True, I could update the code that emits the instances' ports here: https://github.com/llvm/circt/blob/main/lib/EmitVerilog/EmitVerilog.cpp#L1775-L1776. Note that it currently expects a named expression. The relevant wires get emitted and named explicitly here: https://github.com/llvm/circt/blob/main/lib/EmitVerilog/EmitVerilog.cpp#L2094.

That's why I was saying I kind of like treating these modports like other declarations (i.e. wires) here: https://github.com/llvm/circt/pull/302#discussion_r536947451. It lets the emitter for `rtl::instance` work as-is.

If we think it makes sense to keep treating modports as declarations, I think I can change `isValidVerilogCharacter` to only allow `.` when the value is of type `sv:ModportType`. Maybe that's a reasonable approach?

Reply:
I think that's a good solution but I'd want @lattner to weigh in.

Reply:
I agree we don't want to allow periods in identifiers.  I think you have two options here:

1) Force out of line modport projections to have a name like `i_send`, which is ugly or
2) treat the get mod port operation as an inline expression, which emits its base and then adds the ".send".  

I'd recommend #2.

Reply:
Sounds good, I will use option 2.